File: implementation/scipamato/core/sync/src/main/java/ch/difty/scipamato/core/sync/code/HidingInternalsCodeAggregator.java
package ch.difty.scipamato.core.sync.code;

import java.util.*;

import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

/**
 * The {@link HidingInternalsCodeAggregator} has the purpose of
 * <ul>
 * <li>Providing aggregated codes by
 * <ul>
 * <li>Enriching codes with aggregated codes</li>
 * <li>Filtering out internal codes</li>
 * </ul>
 * </li>
 * <li>providing the aggregated
 * <ul>
 * <li>codesPopulation values</li>
 * <li>codesStudyDesign values</li>
 * </ul>
 * </li>
 * </ul>
 *
 * @author u.joss
 */
@Component
@Scope("prototype")
public class HidingInternalsCodeAggregator implements CodeAggregator {

    private static final short CP_ID_1 = 1;
    private static final short CP_ID_2 = 2;

    private static final short CSD_ID_1 = 1;
    private static final short CSD_ID_2 = 2;
    private static final short CSD_ID_3 = 3;

    private final List<String> internalCodes    = new ArrayList<>();
    private final List<String> codes            = new ArrayList<>();
    private final List<Short>  codesPopulation  = new ArrayList<>();
    private final List<Short>  codesStudyDesign = new ArrayList<>();

    @Override
    public void setInternalCodes(final List<String> internalCodes) {
        this.internalCodes.clear();
        this.internalCodes.addAll(internalCodes);
    }

    @Override
    public void load(final String[] codes) {
        clearAll();
        if (codes != null) {
            this.codes.addAll(aggregateCodes(codes));
            this.codesPopulation.addAll(gatherCodesPopulation());
            this.codesStudyDesign.addAll(gatherCodesStudyDesign());
        }
    }

    private void clearAll() {
        this.codes.clear();
        this.codesPopulation.clear();
        this.codesStudyDesign.clear();
    }

    private List<String> aggregateCodes(final String[] codeArray) {
        final List<String> codeList = filterAndEnrich(codeArray);
        Collections.sort(codeList);
        return codeList;
    }

    /**
     * HARDCODED consider moving aggregation into some table in scipamato-core (see
     * also CodeSyncConfig#selectSql)
     */
    private List<String> filterAndEnrich(final String[] codeArray) {
        final Set<String> filtered = new HashSet<>();
        for (final String code : codeArray) {
            if ("5A".equals(code) || "5B".equals(code) || "5C".equals(code))
                filtered.add("5abc");
            else if (!internalCodes.contains(code))
                filtered.add(code);
        }
        return new ArrayList<>(filtered);
    }

    private List<Short> gatherCodesPopulation() {
        final List<Short> pcs = new ArrayList<>();
        if (codes
            .stream()
            .anyMatch(x -> "3A".equals(x) || "3B".equals(x)))
            pcs.add(CP_ID_1);
        if (codes
            .stream()
            .anyMatch("3C"::equals))
            pcs.add(CP_ID_2);
        return pcs;
    }

    @SuppressWarnings("SpellCheckingInspection")
    private List<Short> gatherCodesStudyDesign() {
        final List<Short> csds = new ArrayList<>();
        if (codes
            .stream()
            .anyMatch("5abc"::equals))
            csds.add(CSD_ID_1);
        if (codes
            .stream()
            .anyMatch(x -> "5E".equals(x) || "5F".equals(x) || "5G".equals(x) || "5H".equals(x) || "5I".equals(x)))
            csds.add(CSD_ID_2);
        if (codes
            .stream()
            .anyMatch(x -> "5U".equals(x) || "5M".equals(x)))
            csds.add(CSD_ID_3);
        return csds;
    }

    @Override
    public String[] getAggregatedCodes() {
        return codes.toArray(new String[0]);
    }

    @Override
    public Short[] getCodesPopulation() {
        return codesPopulation.toArray(new Short[0]);
    }

    @Override
    public Short[] getCodesStudyDesign() {
        return codesStudyDesign.toArray(new Short[0]);
    }

}

File: core/core-sync/src/main/kotlin/ch/difty/scipamato/core/sync/code/HidingInternalsCodeAggregator.kt
@file:Suppress("MagicNumber")

package ch.difty.scipamato.core.sync.code

import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component

private const val AGGR_5ABC = "5abc"

/**
 * HARDCODED consider moving aggregation into some table in scipamato-core (see also CodeSyncConfig#selectSql)
 */

private val codeAggregation: Map<String, String> = setOf("5A", "5B", "5C").associateWith { AGGR_5ABC }

private val populationCodeMapper: Map<String, Short> = (
    setOf("3A", "3B").associateWith { 1 } +
        setOf("3C").associateWith { 2 }
    ).map { it.key to it.value.toShort() }.toMap()

private val studyDesignCodeMapper: Map<String, Short> = (
    setOf(AGGR_5ABC).associateWith { 1 } +
        setOf("5E", "5F", "5G", "5H", "5I").associateWith { 2 } +
        setOf("5U", "5M").associateWith { 3 }
    ).map { it.key to it.value.toShort() }.toMap()

/**
 * The [HidingInternalsCodeAggregator] has the purpose of
 *
 *  * Providing aggregated codes by
 *    * Enriching codes with aggregated codes
 *    * Filtering out internal codes
 *
 *  * providing the aggregated
 *    * codesPopulation values
 *    * codesStudyDesign values
 */
@Component
@Scope("prototype")
class HidingInternalsCodeAggregator : CodeAggregator {
    private val _internalCodes: MutableList<String> = ArrayList()
    private val _codes: MutableList<String> = ArrayList()
    private val _codesPopulation: MutableList<Short> = ArrayList()
    private val _codesStudyDesign: MutableList<Short> = ArrayList()

    override fun setInternalCodes(internalCodes: List<String>) {
        this._internalCodes.clear()
        this._internalCodes.addAll(internalCodes)
    }

    override fun load(codes: Array<String>) {
        clearAll()
        this._codes.addAll(aggregateCodes(codes))
        _codesPopulation.addAll(gatherCodesPopulation())
        _codesStudyDesign.addAll(gatherCodesStudyDesign())
    }

    private fun clearAll() {
        _codes.clear()
        _codesPopulation.clear()
        _codesStudyDesign.clear()
    }

    private fun aggregateCodes(codeArray: Array<String>): List<String> =
        filterAndEnrich(codeArray).sorted()

    private fun filterAndEnrich(codeArray: Array<String>): List<String> =
        codeArray
            .map { codeAggregation[it] ?: it }
            .filterNot { it in _internalCodes }
            .distinct()

    private fun gatherCodesPopulation(): List<Short> =
        _codes.mapNotNull { populationCodeMapper[it] }.distinct()

    private fun gatherCodesStudyDesign(): List<Short> =
        _codes.mapNotNull { studyDesignCodeMapper[it] }.distinct()

    override val aggregatedCodes get() = _codes.toTypedArray()

    override val codesPopulation get() = _codesPopulation.toTypedArray()

    override val codesStudyDesign get() = _codesStudyDesign.toTypedArray()
}
